src/io/dataset.cpp

```diff
@@ -324,6 +329,14 @@ void Dataset::Construct(
     max_bin_by_feature_.resize(num_total_features_);
     max_bin_by_feature_.assign(io_config.max_bin_by_feature.begin(), io_config.max_bin_by_feature.end());
   }
+  // get categorical features from the bin types so that we can read the forced bin bounds
+  std::unordered_set<int> categorical_features;
+  for (int i = 0; i < num_total_features_; ++i){
+    if ((bin_mappers[i] != nullptr) && (bin_mappers[i]->bin_type() == BinType::CategoricalBin)){
+      categorical_features.insert(i);
+    }
+  }
+  forced_bin_bounds_ = Dataset::GetForcedBins(io_config.forcedbins_filename, num_total_features_, categorical_features);
```

Comment:
@guolinke Do you mean this part? My understanding is that in this method `bin_mappers` is passed in as an argument and the bin mappers are constructed already. It seems that there are 2 uses of this method (`DatasetLoader::ConstructBinMappersFromTextData` and `DatasetLoader::CostrcutFromSampleData`). In both cases it looks like the bins are found before calling `Dataset::Construct` and passed in to the `Construct` method. Please let me know if I'm misunderstanding.

Reply:
okay, you are right!